[Fetch] Object.prototype shouldn't interfere with fetch streams

ricea opened the pull request:
Test that setting Object.prototype.start to a function that throws
doesn't interfere with fetch operations that create a stream. Similarly,
verify that setting throwing accessors for 'type', 'size', and
'highWaterMark' doesn't cause problems.

See https://github.com/whatwg/fetch/pull/781 for background.

ricea committed:
[Fetch] Object.prototype shouldn't interfere with fetch streams

Test that setting Object.prototype.start to a function that throws
doesn't interfere with fetch operations that create a stream. Similarly,
verify that setting throwing accessors for 'type', 'size', and
'highWaterMark' doesn't cause problems.

See https://github.com/whatwg/fetch/pull/781 for background.

jyasskin commented on fetch/api/basic/stream-safe-creation.any.js:
```diff
@@ -0,0 +1,33 @@
+      throw Error('Object.prototype.start was called');
+    };
+    t.add_cleanup(() => delete Object.prototype.start);
+    await creationCases[creationCase];
```
Should this and the same line below be `creationCases[creationCase]()`?

ricea replied:
This is an important problem! Fixed.

jyasskin commented on fetch/api/basic/stream-safe-creation.any.js:
```diff
@@ -0,0 +1,33 @@
+
+for (creationCase of Object.keys(creationCases)) {
+  promise_test(async t => {
+    Object.prototype.start = () => {
```
Why make this one throw on call, when .size() throws on get()? It seems like .start shouldn't be gotten by the `CreateReadableStream()` version either.

ricea replied:
That was indeed not very useful. I replaced it with a version which errors the stream, which is nice because the side-effect is delayed, so if the implementation is (accidentally?) eating exceptions from the constructor, it will still break if it tries to read from the stream later.

jyasskin commented on fetch/api/basic/stream-safe-creation.any.js:
```diff
@@ -0,0 +1,33 @@
+
+const creationCases = {
+  fetch: async () => fetch(location.href),
+  request: () => new Request(location.href),
```
I think I'm missing where [Request(url)](https://fetch.spec.whatwg.org/#dom-request) constructs a ReadableStream.

ricea replied:
IIRC, it actually does in Blink. But I changed the test case to one which always constructs a ReadableStream by the standard.

ricea committed:
Actually run the various stream construction cases

Also make the Request-based cases construct a stream.

ricea committed:
Remove test for throwing start()

Add a test for accessors which return an invalid value rather than
throwing. Also add a test for a start() function which errors the
stream.

gsnedders commented:
AFAIK, @jyasskin's feedback has been addressed. Can this be re-reviewed/landed as needed?

ricea committed:
Make add_cleanup return a Promise

Also make the test additionally run on shared workers and service
workers.
